File: src/com/mishiranu/instantimage/FetchActivity.java
package com.mishiranu.instantimage;

import java.util.ArrayList;

import android.annotation.TargetApi;
import android.app.Activity;
import android.app.DownloadManager;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.view.ContextMenu;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.view.ContextMenu.ContextMenuInfo;
import android.widget.AdapterView;
import android.widget.GridView;
import android.widget.SearchView;
import android.widget.Toast;

import com.mishiranu.instantimage.util.FileManager;
import com.mishiranu.instantimage.util.Preferences;

public class FetchActivity extends Activity implements ViewTreeObserver.OnGlobalLayoutListener,
		SearchView.OnQueryTextListener, MenuItem.OnActionExpandListener, LoadingDialog.Callback,
		AdapterView.OnItemClickListener
{
	private static final String EXTRA_IMAGE_ITEMS = "imageItems";
	
	private ArrayList<ImageItem> mImageItems;
	private final GridAdapter mGridAdapter = new GridAdapter();

	private SearchView mSearchView;
	private GridView mGridView;
	
	private static final int GRID_SPACING_DP = 4;
	
	private boolean mKeepSearchFocus = false;
	
	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		getActionBar().setDisplayHomeAsUpEnabled(true);
		mSearchView = new SearchView(getActionBar().getThemedContext());
		mSearchView.setOnQueryTextListener(this);
		mKeepSearchFocus = savedInstanceState == null;
		mGridView = new GridView(this);
		mGridView.setId(android.R.id.list);
		mGridView.setOnItemClickListener(this);
		mGridView.setAdapter(mGridAdapter);
		float density = getResources().getDisplayMetrics().density;
		int spacing = (int) (GRID_SPACING_DP * density + 0.5f);
		mGridView.setHorizontalSpacing(spacing);
		mGridView.setVerticalSpacing(spacing);
		mGridView.setPadding(spacing, spacing, spacing, spacing);
		mGridView.setScrollBarStyle(GridView.SCROLLBARS_OUTSIDE_OVERLAY);
		mGridView.setClipToPadding(false);
		mGridView.getViewTreeObserver().addOnGlobalLayoutListener(this);
		registerForContextMenu(mGridView);
		addContentView(mGridView, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
				ViewGroup.LayoutParams.MATCH_PARENT));
		LoadingDialog loadingDialog = (LoadingDialog) getFragmentManager().findFragmentByTag(LoadingDialog.TAG);
		if (loadingDialog != null) loadingDialog.setCallback(this);
		if (savedInstanceState != null)
		{
			ArrayList<ImageItem> imageItems = savedInstanceState.getParcelableArrayList(EXTRA_IMAGE_ITEMS);
			if (imageItems != null)
			{
				mImageItems = imageItems;
				mGridAdapter.setImageItems(imageItems);
			}
		}
	}
	
	@Override
	protected void onSaveInstanceState(Bundle outState)
	{
		super.onSaveInstanceState(outState);
		outState.putParcelableArrayList(EXTRA_IMAGE_ITEMS, mImageItems);
	}
	
	private final Runnable mLayoutRunnable = new Runnable()
	{
		@Override
		public void run()
		{
			int position = mGridView.getFirstVisiblePosition();
			float density = getResources().getDisplayMetrics().density;
			int sizeDp = getResources().getConfiguration().smallestScreenWidthDp >= 600 ? 200 : 150;
			int size = (int) (sizeDp * density + 0.5f);
			int spacing = mGridView.getHorizontalSpacing();
			int numColumns = (mLastGridViewWidth - spacing) / (size + spacing);
			mGridView.setNumColumns(numColumns);
			mGridView.setSelection(position);
		}
	};
	
	private int mLastGridViewWidth = -1;
	
	@Override
	public void onGlobalLayout()
	{
		int width = mGridView.getWidth();
		if (mLastGridViewWidth != width)
		{
			mLastGridViewWidth = width;
			mGridView.removeCallbacks(mLayoutRunnable);
			mGridView.post(mLayoutRunnable);
		}
	}
	
	@Override
	public boolean onQueryTextChange(String newText)
	{
		return false;
	}
	
	@Override
	public boolean onQueryTextSubmit(String query)
	{
		mSearchView.clearFocus();
		if (query.length() > 0) new LoadingDialog(query, null, this).show(getFragmentManager(), LoadingDialog.TAG);
		return false;
	}
	
	@Override
	public boolean onMenuItemActionExpand(MenuItem item)
	{
		return true;
	}
	
	@Override
	public boolean onMenuItemActionCollapse(MenuItem item)
	{
		finish();
		return false;
	}
	
	@Override
	public boolean onCreateOptionsMenu(Menu menu)
	{
		menu.add(0, 1, 0, "").setActionView(mSearchView).setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS
				| MenuItem.SHOW_AS_ACTION_COLLAPSE_ACTION_VIEW).setOnActionExpandListener(this).expandActionView();
		menu.add(0, 2, 0, R.string.text_crop_thumbnails).setCheckable(true)
				.setChecked(Preferences.isCropThumbnails(this));
		if (mKeepSearchFocus) mKeepSearchFocus = false; else mSearchView.clearFocus();
		return true;
	}
	
	@Override
	public boolean onOptionsItemSelected(MenuItem item)
	{
		switch (item.getItemId())
		{
			case android.R.id.home:
			{
				finish();
				return true;
			}
			case 2:
			{
				Preferences.setCropThumbnails(this, !item.isChecked());
				mGridAdapter.notifyDataSetChanged();
				invalidateOptionsMenu();
				return true;
			}
		}
		return super.onOptionsItemSelected(item);
	}
	
	@Override
	public void onCreateContextMenu(ContextMenu menu, View v, ContextMenuInfo menuInfo)
	{
		menu.add(0, 1, 0, R.string.text_download);
	}
	
	@TargetApi(Build.VERSION_CODES.M)
	@Override
	public boolean onContextItemSelected(MenuItem item)
	{
		AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) item.getMenuInfo();
		switch (item.getItemId())
		{
			case 1:
			{
				if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
				{
					if (checkSelfPermission(android.Manifest.permission.WRITE_EXTERNAL_STORAGE)
							!= PackageManager.PERMISSION_GRANTED)
					{
						requestPermissions(new String[] {android.Manifest.permission.WRITE_EXTERNAL_STORAGE},
								info.position);
						return true;
					}
				}
				ImageItem imageItem = mGridAdapter.getItem(info.position);
				performDownload(imageItem);
				return true;
			}
		}
		return false;
	}
	
	@Override
	public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults)
	{
		if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED)
		{
			performDownload(mGridAdapter.getItem(requestCode));
		}
	}
	
	private void performDownload(ImageItem imageItem)
	{
		DownloadManager downloadManager = (DownloadManager) getSystemService(DOWNLOAD_SERVICE);
		DownloadManager.Request request = new DownloadManager.Request(Uri.parse(imageItem.imageUriString));
		request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);
		request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS,
				FileManager.obtainSimpleFileName(imageItem.imageUriString));
		request.allowScanningByMediaScanner();
		try
		{
			downloadManager.enqueue(request);
		}
		catch (IllegalArgumentException e)
		{
			String message = e.getMessage();
			if (message.equals("Unknown URL content://downloads/my_downloads"))
			{
				Toast.makeText(this, R.string.error_download_manager, Toast.LENGTH_LONG).show();
			}
			else throw e;
		}
	}
	
	private final Runnable mPositionRunnable = new Runnable()
	{
		@Override
		public void run()
		{
			mGridView.setSelection(0);
		}
	};
	
	@Override
	public void onQueryLoadingSuccess(ArrayList<ImageItem> imageItems)
	{
		mImageItems = imageItems;
		mGridAdapter.setImageItems(imageItems);
		mGridView.post(mPositionRunnable);
	}
	
	@Override
	public void onQueryLoadingError(int errorMessageId)
	{
		Toast.makeText(this, getString(errorMessageId), Toast.LENGTH_SHORT).show();
	}
	
	@Override
	public void onImageLoadingSuccess(int contentId, String mimeType)
	{
		Uri uri = ImageProvider.build(contentId);
		setResult(RESULT_OK, new Intent().setDataAndType(uri, mimeType));
		finish();
	}
	
	@Override
	public void onImageLoadingError(int errorMessageId)
	{
		onQueryLoadingError(errorMessageId);
	}
	
	public void onItemClick(AdapterView<?> parent, View view, int position, long id)
	{
		ImageItem imageItem = mGridAdapter.getItem(position);
		new LoadingDialog(null, imageItem.imageUriString, this).show(getFragmentManager(), LoadingDialog.TAG);
	}
}

File: src/com/mishiranu/instantimage/ui/FetchActivity.scala
package com.mishiranu.instantimage.ui

import android.app.{Activity, ActivityOptions}
import android.content.{ClipData, Intent}
import android.graphics.drawable.Drawable
import android.os.{Bundle, Parcelable}
import android.view.{Menu, MenuItem, View, ViewGroup, ViewTreeObserver}
import android.widget.{AdapterView, GridView, SearchView, Toast}

import com.mishiranu.instantimage.R
import com.mishiranu.instantimage.model.Image
import com.mishiranu.instantimage.util.Preferences
import com.mishiranu.instantimage.util.ScalaHelpers._

import scala.collection.mutable

class FetchActivity extends Activity with ViewTreeObserver.OnGlobalLayoutListener with SearchView.OnQueryTextListener
  with MenuItem.OnActionExpandListener with LoadingDialog.Callback with AdapterView.OnItemClickListener
  with AdapterView.OnItemLongClickListener {
  import FetchActivity._

  private val images = new mutable.ArrayBuffer[Image]
  private val selectedImages = new mutable.LinkedHashSet[Image]
  private val gridAdapter = new GridAdapter()

  private var searchView: SearchView = _
  private var gridView: GridView = _

  private var keepSearchFocus: Boolean = false

  protected override def onCreate(savedInstanceState: Bundle): Unit = {
    super.onCreate(savedInstanceState)
    getActionBar.setDisplayHomeAsUpEnabled(true)

    searchView = new SearchView(getActionBar.getThemedContext)
    searchView.setOnQueryTextListener(this)

    keepSearchFocus = savedInstanceState == null

    gridView = new GridView(this)
    gridView.setId(android.R.id.list)
    gridView.setOnItemClickListener(this)
    if (getIntent.getBooleanExtra(Intent.EXTRA_ALLOW_MULTIPLE, false)) {
      gridView.setOnItemLongClickListener(this)
    }
    gridView.setAdapter(gridAdapter)
    val density = getResources.getDisplayMetrics.density
    val spacing = (GRID_SPACING_DP * density + 0.5f).toInt
    gridView.setHorizontalSpacing(spacing)
    gridView.setVerticalSpacing(spacing)
    gridView.setPadding(spacing, spacing, spacing, spacing)
    gridView.setScrollBarStyle(View.SCROLLBARS_OUTSIDE_OVERLAY)
    gridView.setClipToPadding(false)
    gridView.getViewTreeObserver.addOnGlobalLayoutListener(this)
    addContentView(gridView, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
        ViewGroup.LayoutParams.MATCH_PARENT))

    val loadingDialog = getFragmentManager.findFragmentByTag(LoadingDialog.TAG).asInstanceOf[LoadingDialog]
    if (loadingDialog != null) {
      loadingDialog.setCallback(this)
    }
    if (savedInstanceState != null) {
      val images = savedInstanceState.getParcelableArrayList[Image](EXTRA_IMAGES)
      if (images != null) {
        import scala.collection.JavaConversions._
        this.images ++= images
        gridAdapter.setImages(this.images)
      }
      val selectedImages = savedInstanceState.getParcelableArrayList[Image](EXTRA_SELECTED_IMAGES)
      if (selectedImages != null) {
        import scala.collection.JavaConversions._
        this.selectedImages ++= selectedImages
      }
    }
  }

  override protected def onSaveInstanceState(outState: Bundle): Unit = {
    super.onSaveInstanceState(outState)
    import scala.collection.JavaConverters._
    outState.putParcelableArrayList(EXTRA_IMAGES, new java.util.ArrayList(images.asJava))
    outState.putParcelableArrayList(EXTRA_SELECTED_IMAGES, new java.util.ArrayList(selectedImages.asJava))
  }

  private var lastGridViewWidth = -1

  private val layoutRunnable: Runnable = Runnable {
    val position = gridView.getFirstVisiblePosition
    val density = getResources.getDisplayMetrics.density
    val sizeDp = if (getResources.getConfiguration.smallestScreenWidthDp >= 600) 200 else 150
    val size = (sizeDp * density + 0.5f).toInt
    val spacing = gridView.getHorizontalSpacing
    val numColumns = (lastGridViewWidth - spacing) / (size + spacing)
    gridView.setNumColumns(numColumns)
    gridView.setSelection(position)
  }

  override def onGlobalLayout(): Unit = {
    val width = gridView.getWidth
    if (lastGridViewWidth != width) {
      lastGridViewWidth = width
      gridView.removeCallbacks(layoutRunnable)
      gridView.post(layoutRunnable)
    }
  }

  override def onQueryTextChange(newText: String): Boolean = false

  override def onQueryTextSubmit(query: String): Boolean = {
    searchView.clearFocus()
    if (query.length() > 0) {
      new LoadingDialog(query, this).show(getFragmentManager, LoadingDialog.TAG)
    }
    false
  }

  override def onMenuItemActionExpand(item: MenuItem): Boolean = true

  override def onMenuItemActionCollapse(item: MenuItem): Boolean = {
    finish()
    false
  }

  override def onCreateOptionsMenu(menu: Menu): Boolean = {
    // noinspection ScalaDeprecation
    def getDrawable(resId: Int): Drawable = sdk(21) {
      case true => FetchActivity.this.getDrawable(resId)
      case false => getResources.getDrawable(resId)
    }.get

    menu.add(0, OPTIONS_ITEM_SEARCH, 0, "").setActionView(searchView)
      .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS | MenuItem.SHOW_AS_ACTION_COLLAPSE_ACTION_VIEW)
      .setOnActionExpandListener(this).expandActionView()
    menu.add(0, OPTIONS_ITEM_CROP, 0, R.string.text_crop_thumbnails).setIcon(getDrawable(sdk(21) {
      case true => R.drawable.ic_crop_white
      case false => R.drawable.ic_action_crop
    }.get).mutate).setCheckable(true).setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
    if (keepSearchFocus) {
      keepSearchFocus = false
    } else {
      searchView.clearFocus()
    }
    true
  }

  override def onPrepareOptionsMenu(menu: Menu): Boolean = {
    val cropMenuItem = menu.findItem(OPTIONS_ITEM_CROP)
    cropMenuItem.setChecked(Preferences.cropThumbnails)
    cropMenuItem.getIcon.setAlpha(if (Preferences.cropThumbnails) 0xff else 0x7f)
    super.onPrepareOptionsMenu(menu)
  }

  override def onOptionsItemSelected(item: MenuItem): Boolean = {
    item.getItemId match {
      case android.R.id.home =>
        finish()
        true
      case OPTIONS_ITEM_CROP =>
        Preferences.setCropThumbnails(!item.isChecked)
        gridAdapter.notifyDataSetChanged()
        invalidateOptionsMenu()
        true
      case _ =>
        super.onOptionsItemSelected(item)
    }
  }

  override def onActivityResult(requestCode: Int, resultCode: Int, data: Intent): Unit = {
    requestCode match {
      case REQUEST_CODE_PREVIEW =>
        if (resultCode == Activity.RESULT_OK) {
          val image = data.getParcelableExtra[Image](PreviewActivity.EXTRA_IMAGE)
          if (!selectedImages.contains(image)) {
            val position = images.indexOf(image)
            val view = gridView.getChildAt(position - gridView.getFirstVisiblePosition)
            toggleSelectedItem(image, view)
          }
          selectedImages += image
          new LoadingDialog(selectedImages.map(_.imageUriString), this).show(getFragmentManager, LoadingDialog.TAG)
        }
      case _ =>
        super.onActivityResult(requestCode, resultCode, data)
    }
  }

  override def onQueryLoadingSuccess(images: List[Image]): Unit = {
    this.images.clear()
    this.selectedImages.clear()
    this.images ++= images
    gridAdapter.setImages(images)
    gridView.post(gridView.setSelection(0))
  }

  override def onQueryLoadingError(errorMessageId: Int): Unit = {
    Toast.makeText(this, getString(errorMessageId), Toast.LENGTH_SHORT).show()
  }

  override def onImageLoadingSuccess(results: List[LoadingDialog.ImageLoadingSuccessResult]): Unit = {
    val intent = new Intent
    if (results.length >= 2) {
      intent.setClipData {
        val mimeTypes = if (getIntent.hasExtra(Intent.EXTRA_MIME_TYPES)) {
          getIntent.getStringArrayExtra(Intent.EXTRA_MIME_TYPES)
        } else {
          Array(getIntent.getType)
        }

        import scala.language.implicitConversions
        implicit def obtainClipDataItem(result: LoadingDialog.ImageLoadingSuccessResult): ClipData.Item = {
          new ClipData.Item(ImageProvider.buildUri(result.contentId))
        }

        val clipData = new ClipData(null, mimeTypes, results.head)
        for (i <- 1 until results.length) {
          clipData.addItem(results(i))
        }
        clipData
      }
    } else {
      val result = results.head
      val uri = ImageProvider.buildUri(result.contentId)
      intent.setDataAndType(uri, result.mimeType)
    }
    setResult(Activity.RESULT_OK, intent)
    finish()
  }

  override def onImageLoadingError(errorMessageId: Int): Unit = {
    onQueryLoadingError(errorMessageId)
  }

  override def onItemClick(parent: AdapterView[_], view: View, position: Int, id: Long): Unit = {
    val image = gridAdapter.getItem(position)
    val view = parent.getChildAt(position - parent.getFirstVisiblePosition)
    if (selectedImages.contains(image)) {
      toggleSelectedItem(image, view)
    } else {
      val intent = new Intent(this, classOf[PreviewActivity])
        .putExtra(PreviewActivity.EXTRA_IMAGE, image.asInstanceOf[Parcelable])
        .putExtra(PreviewActivity.EXTRA_SELECTED_COUNT, selectedImages.size + 1)
      sdk(21) {
        case true =>
          val (imageView, transition) = gridAdapter.extractImageViewForTransition(view)
          intent.putExtra(PreviewActivity.EXTRA_TRANSITION, transition)
          startActivityForResult(intent, REQUEST_CODE_PREVIEW,
            ActivityOptions.makeSceneTransitionAnimation(this, imageView -> transition).toBundle)
        case false =>
          startActivityForResult(intent, REQUEST_CODE_PREVIEW)
      }
    }
  }

  override def onItemLongClick(parent: AdapterView[_], view: View, position: Int, id: Long): Boolean = {
    val image = gridAdapter.getItem(position)
    val view = parent.getChildAt(position - parent.getFirstVisiblePosition)
    toggleSelectedItem(image, view)
    true
  }

  private def toggleSelectedItem(image: Image, view: View): Unit = {
    val selected = if (selectedImages.contains(image)) {
      selectedImages -= image
      false
    } else {
      selectedImages += image
      true
    }
    gridAdapter.setSelected(view, image, selected)
  }
}

object FetchActivity {
  private val EXTRA_IMAGES = "images"
  private val EXTRA_SELECTED_IMAGES = "selected_images"

  private val REQUEST_CODE_PREVIEW = 1

  private val OPTIONS_ITEM_SEARCH = 1
  private val OPTIONS_ITEM_CROP = 2

  private val GRID_SPACING_DP = 4
}
